doc/proxy-settings.md

# ProxySettings

Represents the proxy server configurations for API calls.

## Properties

| Name | Type | Tag | Description |
|  --- | --- | --- | --- |
| address | `str` | required | The proxy server URL. |
| port | `int` | optional | The port to connect to the proxy server. |
| username | `str` | optional | Username for proxy authentication. |
| password | `str` | optional | Password for proxy authentication. |

## Usage Example

```python
from frankfurterapi.frankfurterapi_client import FrankfurterapiClient
from frankfurterapi.http.proxy_settings import ProxySettings

client = FrankfurterapiClient(
    proxy_settings=ProxySettings(
        address='http://localhost',
        port=8888,
        username='user',
        password='pass'
     )
)
```



doc/environment-based-client-initialization.md

# Environment-Based Client Initialization

The SDK client can also be initialized directly from environment variables using the `from_environment()` class method. This allows the SDK to automatically read configuration values from the runtime environment or a .env file.

## Example

```python
from frankfurterapi.frankfurterapi_client import FrankfurterapiClient

# Specify the path to your .env file if it’s located outside the project’s root directory.
client = FrankfurterapiClient.from_environment(dotenv_path='/path/to/.env')
```

You can also specify a path to a `.env` file by passing it to the `from_environment()` method:

The same method can accept keyword arguments to override any values read from the environment, and the arguments to override values should follow the same approach as code-based client initialization.

```python
from frankfurterapi.frankfurterapi_client import FrankfurterapiClient

client = FrankfurterapiClient.from_environment(
    dotenv_path='/path/to/.env',
    timeout=0,  # overrides timeout from environment variable
)
```

Values provided through arguments take precedence over those defined in environment variables.

## Example `.env` File

```python
ENVIRONMENT=production


TIMEOUT=60
MAX_RETRIES=3
BACKOFF_FACTOR=2
RETRY_STATUSES=408,413
RETRY_METHODS=GET,PUT,DELETE

# Logging Configuration
LOG_LEVEL=DEBUG
MASK_SENSITIVE_HEADERS=true

# Request Logging Configuration
REQUEST_LOG_BODY=true
REQUEST_LOG_HEADERS=true
REQUEST_INCLUDE_QUERY_IN_PATH=true
REQUEST_HEADERS_TO_INCLUDE=Content-Type,X-Request-ID
REQUEST_HEADERS_TO_EXCLUDE=Authorization
REQUEST_HEADERS_TO_UNMASK=X-Request-ID

# Response Logging Configuration
RESPONSE_LOG_BODY=true
RESPONSE_LOG_HEADERS=true
RESPONSE_HEADERS_TO_INCLUDE=Content-Type,X-Correlation-ID,Date,Server
RESPONSE_HEADERS_TO_EXCLUDE=Set-Cookie,Authorization,X-API-Key
RESPONSE_HEADERS_TO_UNMASK=X-Correlation-ID

# Proxy Configuration
PROXY_ADDRESS=http://localhost:3000
PROXY_PORT=8080
PROXY_USERNAME=username
PROXY_PASSWORD=password
```

## Note

- If an environment variable is not defined, the default SDK configuration value will be used.



frankfurterapi/configuration.py
"""frankfurterapi.

This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
"""

# ruff: noqa: E501
import os
from enum import Enum

from apimatic_core.http.configurations.http_client_configuration import (
    HttpClientConfiguration,
)
from apimatic_requests_client_adapter.requests_client import (
    RequestsClient,
)
from dotenv import load_dotenv

from frankfurterapi.http.proxy_settings import (
    ProxySettings,
)
from frankfurterapi.logging.configuration.api_logging_configuration import (
    LoggingConfiguration,
)


class Environment(Enum):
    """An enum for SDK environments."""

    PRODUCTION = 0

    @classmethod
    def from_value(cls, value, default=None):
        """Convert a value (string or int) to an Environment enum member.

        Args:
            value (Union[str, int]): The value to convert.
            default (Environment): The fallback enum member if conversion fails.

        Returns:
            Environment: Matching enum member or fallback if invalid.

        """
        if value is None:
            return default

        # Try to match directly by enum member
        if isinstance(value, cls):
            return value

        # Handle integer or string conversion
        for member in cls:
            if (str(member.value).lower() == str(value).lower()
                or member.name.lower() == str(value).lower()):
                return member

        # Fallback to provided default
        return default


class Server(Enum):
    """An enum for API servers."""

    DEFAULT = 0

    @classmethod
    def from_value(cls, value, default=None):
        """Convert a value (string or int) to a Server enum member.

        Args:
            value (Union[str, int]): The value to convert.
            default (Server): The fallback enum member if conversion fails.

        Returns:
            Server: Matching enum member or fallback if invalid.

        """
        if value is None:
            return default

        # Try to match directly by enum member
        if isinstance(value, cls):
            return value

        # Handle integer or string conversion
        for member in cls:
            if (str(member.value).lower() == str(value).lower()
                or member.name.lower() == str(value).lower()):
                return member

        # Fallback to provided default
        return default


class Configuration(HttpClientConfiguration):
    """A class used for configuring the SDK by a user."""

    @property
    def environment(self):
        """Return current environment."""
        return self._environment

    def __init__(self, http_client_instance=None,
                 override_http_client_configuration=False, http_call_back=None,
                 timeout=30, max_retries=0, backoff_factor=2, retry_statuses=None,
                 retry_methods=None, proxy_settings=None, logging_configuration=None,
                 environment=Environment.PRODUCTION):
        """Initialize Configuration object."""
        if retry_methods is None:
            retry_methods = ["GET", "PUT", "GET", "PUT"]

        if retry_statuses is None:
            retry_statuses = [408, 413, 429, 500, 502, 503, 504, 521,
                522, 524, 408, 413, 429, 500, 502, 503, 504, 521, 522, 524]

        super().__init__(
            http_client_instance=http_client_instance,
            override_http_client_configuration=override_http_client_configuration,
            http_call_back=http_call_back, timeout=timeout,
            max_retries=max_retries, backoff_factor=backoff_factor,
            retry_statuses=retry_statuses, retry_methods=retry_methods,
            proxy_settings=proxy_settings,
            logging_configuration=logging_configuration,
        )

        # Current API environment
        self._environment = environment

        # The Http Client to use for making requests.
        self.set_http_client(self.create_http_client())

    def clone_with(self, http_client_instance=None,
                   override_http_client_configuration=None, http_call_back=None,
                   timeout=None, max_retries=None, backoff_factor=None,
                   retry_statuses=None, retry_methods=None, proxy_settings=None,
                   logging_configuration=None, environment=None):
        """Clone configuration with overrides."""
        http_client_instance = http_client_instance or self.http_client_instance
        override_http_client_configuration =\
            (override_http_client_configuration
            or self.override_http_client_configuration)
        http_call_back = http_call_back or self.http_callback
        timeout = timeout or self.timeout
        max_retries = max_retries or self.max_retries
        backoff_factor = backoff_factor or self.backoff_factor
        retry_statuses = retry_statuses or self.retry_statuses
        retry_methods = retry_methods or self.retry_methods
        proxy_settings = proxy_settings or self.proxy_settings
        logging_configuration = logging_configuration or self.logging_configuration
        environment = environment or self.environment
        return Configuration(
            http_client_instance=http_client_instance,
            override_http_client_configuration=override_http_client_configuration,
            http_call_back=http_call_back, timeout=timeout, max_retries=max_retries,
            backoff_factor=backoff_factor, retry_statuses=retry_statuses,
            retry_methods=retry_methods, proxy_settings=proxy_settings,
            logging_configuration=logging_configuration, environment=environment,
        )

    def create_http_client(self):
        """Create the HTTP client instance."""
        return RequestsClient(
            timeout=self.timeout, max_retries=self.max_retries,
            backoff_factor=self.backoff_factor, retry_statuses=self.retry_statuses,
            retry_methods=self.retry_methods,
            http_client_instance=self.http_client_instance,
            override_http_client_configuration=self.override_http_client_configuration,
            response_factory=self.http_response_factory,
            proxies=self.proxy_settings.to_proxies() if self.proxy_settings else None,
        )

    # All the environments the SDK can run in
    environments = {
        Environment.PRODUCTION: {
            Server.DEFAULT: "https://api.frankfurter.dev/v2",
        },
    }

    def get_base_uri(self, server=Server.DEFAULT):
        """Generate the appropriate base URI for the environment and the server.

        Args:
            server (Configuration.Server): The server enum for which the base
            URI is required.

        Returns:
            String: The base URI.

        """
        return self.environments[self.environment][server]

    @classmethod
    def from_environment(cls, dotenv_path=None, **overrides):
        """Create Configuration object from .env and environment variables.

        Args:
            dotenv_path (str, optional): Path to the .env file to load.
             If None, the default .env file is used.
            **overrides: Optional values overriding setting from environment variables.

        Returns:
            Configuration: A configuration object populated with the resolved values.

        """
        # load .env automatically
        load_dotenv(dotenv_path or None, override=True)

        override_http_client_configuration = os.getenv(
            "OVERRIDE_HTTP_CLIENT_CONFIGURATION", "false").lower() == "true"
        timeout = int(os.getenv("TIMEOUT", "30"))
        max_retries = int(os.getenv("MAX_RETRIES", "0"))
        backoff_factor = float(os.getenv("BACKOFF_FACTOR", "2"))
        statuses = os.getenv("RETRY_STATUSES", None)
        retry_statuses = [int(v.strip()) for v in statuses.split(",")
                          if v.strip().isdigit()] if statuses else None
        methods = os.getenv("RETRY_METHODS", None)
        retry_methods = [v.strip() for v in methods.split(",") if v.strip()]\
            if methods else None
        environment = Environment.from_value(
            os.getenv("ENVIRONMENT"), Environment.PRODUCTION)

        # Preparing default configuration
        default_config = cls(
            override_http_client_configuration=override_http_client_configuration,
            timeout=timeout,
            max_retries=max_retries,
            backoff_factor=backoff_factor,
            retry_statuses=retry_statuses,
            retry_methods=retry_methods,
            environment=environment,
            proxy_settings=ProxySettings.from_environment(),
            logging_configuration=LoggingConfiguration.from_environment(),
        )

        return default_config.clone_with(**overrides)
